File: api/webhook/stripe.py
from http.server import BaseHTTPRequestHandler
import os
import json
import stripe

stripe.api_key = os.environ.get("STRIPE_SECRET_KEY")
webhook_secret = os.environ.get("STRIPE_WEBHOOK_SECRET")

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)
        sig_header = self.headers.get('Stripe-Signature')

        try:
            event = stripe.Webhook.construct_event(
                post_data, sig_header, webhook_secret
            )
        except Exception as e:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(e)}).encode())
            return

        if event['type'] == 'checkout.session.completed':
            session = event['data']['object']
            customer_email = session.get('customer_details', {}).get('email')
            # Lógica de liberação aqui

        self.send_response(200)
        self.end_headers()
        self.wfile.write(json.dumps({"received": True}).encode())

File: api/webhook/stripe.ts
import { NextResponse } from 'next/server';
import Stripe from 'stripe';

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!, {
  apiVersion: '2023-10-16' as any,
});

const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET!;

export async function POST(req: Request) {
  const body = await req.text();
  const signature = req.headers.get('stripe-signature') as string;

  let event: Stripe.Event;

  try {
    event = stripe.webhooks.constructEvent(body, signature, webhookSecret);
  } catch (err: any) {
    console.error(`Erro na assinatura do Webhook: ${err.message}`);
    return NextResponse.json({ error: `Webhook Error: ${err.message}` }, { status: 400 });
  }

  // Trata o pagamento concluído
  if (event.type === 'checkout.session.completed') {
    const session = event.data.object as Stripe.Checkout.Session;
    const customerEmail = session.customer_details?.email;
    const subscriptionId = session.subscription;

    // TODO: Liberar o acesso do usuário
    // Exemplo: Atualizar no banco de dados que o email possui plano ativo
    console.log(`Pagamento aprovado para: ${customerEmail}`);
  }

  return NextResponse.json({ received: true }, { status: 200 });
}
